File: android/data/src/main/kotlin/com/zandaulion/omaha/data/Dcf.kt
package com.zandaulion.omaha.data

import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToInt

/**
 * The DCF sandbox model, in Kotlin.
 *
 * **This is the project's only deliberate second implementation**, and it needs
 * its justification stated where someone will find it.
 *
 * Everything else that both clients compute lives in `core/` and runs through
 * QuickJS, because a second implementation is a second set of answers. The
 * sandbox is the exception for one reason: it recomputes on every slider frame.
 * A QuickJS call costs a fresh interpreter and a bundle parse — about 21 ms per
 * call under the alpha13 workaround (doc 13 §24) — which a drag cannot afford.
 *
 * `core/analysis/dcf.js` remains the definition. `scripts/gen-dcf-vectors.mjs`
 * records its output into `core/__fixtures__/dcf.vectors.json`, and
 * `DcfParityTest` asserts this file agrees with those numbers. A change to the
 * model that is not mirrored here fails the Android build rather than shipping
 * two fair values for one company.
 */
object Dcf {

    const val PROJECTION_YEARS = 5

    data class Range(val min: Double, val max: Double)

    val GROWTH_PCT = Range(-25.0, 45.0)
    val MULTIPLE = Range(8.0, 45.0)
    val DISCOUNT_PCT = Range(6.0, 16.0)

    data class Assumptions(val growthPct: Int, val multiple: Double, val discountPct: Double)

    data class Year(val year: Int, val fcf: Double, val pv: Double)

    data class Projection(
        val rows: List<Year>,
        val cumulativePV: Double,
        val terminalValue: Double,
        val pvTerminal: Double,
        val enterpriseValue: Double,
        val equityValue: Double,
        val fairValue: Double
    )

    enum class VerdictKind { NoEquityValue, NoPrice, Divergent, Undervalued, Overvalued }

    data class Verdict(val kind: VerdictKind, val pct: Double?, val factor: Double?)

    /** Where the sliders start: what the scorecard was actually built from. */
    fun baselines(growthRate: Double?, terminalMultiple: Double?, discountRate: Double?): Assumptions =
        clamp(
            Assumptions(
                growthPct = ((growthRate ?: 0.06) * 100).roundToInt(),
                multiple = (terminalMultiple ?: 15.0).roundToInt().toDouble(),
                discountPct = ((discountRate ?: 0.095) * 1000).roundToInt() / 10.0
            )
        )

    /**
     * Held inside the slider's range at **both** ends.
     *
     * Clamping one end only was a real defect in the original: a company whose
     * filed growth sat outside the range produced a preset the slider could not
     * represent, so the control pinned at its limit while the fair value below
     * used the unclamped figure. The two disagreed and nothing said so.
     */
    fun clamp(a: Assumptions) = Assumptions(
        growthPct = min(GROWTH_PCT.max, max(GROWTH_PCT.min, a.growthPct.toDouble())).toInt(),
        multiple = min(MULTIPLE.max, max(MULTIPLE.min, a.multiple)),
        discountPct = min(DISCOUNT_PCT.max, max(DISCOUNT_PCT.min, a.discountPct))
    )

    /**
     * Bear is 35% *worse* rather than 35% lower.
     *
     * Not the same thing for a shrinking business: multiplying a negative
     * growth rate by 0.65 would make the bear case the optimistic one.
     */
    fun preset(name: String, baselines: Assumptions): Assumptions {
        val g = baselines.growthPct
        return when (name) {
            "bear" -> clamp(
                Assumptions(
                    growthPct = (if (g >= 0) g * 0.65 else g * 1.35).roundToInt(),
                    multiple = 16.0,
                    discountPct = 11.0
                )
            )
            "bull" -> clamp(
                Assumptions(
                    growthPct = (if (g >= 0) g * 1.30 else g * 0.70).roundToInt(),
                    multiple = 32.0,
                    discountPct = 9.0
                )
            )
            else -> clamp(baselines)
        }
    }

    /** Why the model cannot run, or null. Never a fabricated cash flow. */
    fun blockedReason(applicable: Boolean?, reason: String?, cashFlowBase: Double?, shares: Double?): String? {
        if (applicable == false) return reason ?: "not-applicable"
        if (cashFlowBase == null || !cashFlowBase.isFinite() || cashFlowBase <= 0) return "negative-fcf"
        if (shares == null || !shares.isFinite() || shares <= 0) return "no-share-count"
        return null
    }

    fun explainBlocked(reason: String): String = when (reason) {
        "negative-fcf" ->
            "This company is not generating positive free cash flow, so a discounted cash " +
                "flow model has nothing to discount. Judge it on the balance sheet and the " +
                "path back to cash generation instead."
        "no-share-count" ->
            "The diluted share count is not in the filings for this listing, so a per-share " +
                "value cannot be derived."
        "not-meaningful-for-financials" ->
            "Free cash flow is not owner earnings for a bank or insurer — deposit and loan " +
                "flows dominate it. Book value and return on equity are the measures that " +
                "apply here."
        else -> "This model cannot be run on the available filings."
    }

    fun project(
        cashFlowBase: Double,
        shares: Double,
        netCash: Double,
        growthPct: Int,
        multiple: Double,
        discountPct: Double
    ): Projection {
        val g = growthPct / 100.0
        val r = discountPct / 100.0

        val rows = ArrayList<Year>(PROJECTION_YEARS)
        var fcf = cashFlowBase
        var cumulativePV = 0.0

        for (t in 1..PROJECTION_YEARS) {
            fcf *= (1 + g)
            val pv = fcf / (1 + r).pow(t)
            cumulativePV += pv
            rows.add(Year(t, fcf, pv))
        }

        val terminalValue = fcf * multiple
        val pvTerminal = terminalValue / (1 + r).pow(PROJECTION_YEARS)
        val enterpriseValue = cumulativePV + pvTerminal
        val equityValue = enterpriseValue + netCash

        return Projection(
            rows = rows,
            cumulativePV = cumulativePV,
            terminalValue = terminalValue,
            pvTerminal = pvTerminal,
            enterpriseValue = enterpriseValue,
            equityValue = equityValue,
            fairValue = equityValue / shares
        )
    }

    /**
     * What the result means against the traded price.
     *
     * `Divergent` exists because a fair value several multiples from the price
     * almost always means the assumptions are wrong, and presenting that as an
     * enormous margin of safety invites exactly the wrong conclusion.
     * `Overvalued` is a premium over fair value rather than a negative margin,
     * because the margin form reaches −188% on an expensive stock and stops
     * carrying meaning.
     */
    fun verdict(fairValue: Double, price: Double): Verdict {
        if (!fairValue.isFinite() || fairValue <= 0) return Verdict(VerdictKind.NoEquityValue, null, null)
        if (!price.isFinite() || price <= 0) return Verdict(VerdictKind.NoPrice, null, null)

        val factor = fairValue / price
        if (factor >= 3 || factor <= 1.0 / 3.0) return Verdict(VerdictKind.Divergent, null, factor)
        return if (fairValue > price) {
            Verdict(VerdictKind.Undervalued, (fairValue - price) / fairValue * 100, factor)
        } else {
            Verdict(VerdictKind.Overvalued, (price - fairValue) / fairValue * 100, factor)
        }
    }
}

File: core/analysis/dcf.js
/**
 * The DCF sandbox model.
 *
 * Pure, and deliberately here rather than in a client. This arithmetic lived
 * only in `web/app.js`, which was fine while there was one client and becomes
 * a drift generator the moment there are two — the sliders recompute on every
 * drag, so a second implementation would be a second set of fair values, in the
 * place where the app is at its most advice-shaped.
 *
 * The scoring engine already models a DCF (`core/scoring.js`); that one produces
 * the baseline assumptions and the pillar score. This module is the *sandbox*:
 * the same projection run against assumptions a person chose, so they can ask
 * what would have to be true.
 *
 * Nothing here formats. Callers own their own number rendering, since the two
 * clients disagree about currency symbols and neither should learn the other's.
 */

/** Years projected explicitly before the terminal multiple takes over. */
export const PROJECTION_YEARS = 5;

/** Slider bounds, shared so the two clients cannot offer different ranges. */
export const LIMITS = {
  growthPct: { min: -25, max: 45 },
  multiple: { min: 8, max: 45 },
  discountPct: { min: 6, max: 16 }
};

/**
 * Where the sliders start: exactly what the scorecard was built from.
 *
 * Opening on the engine's own assumptions means the first number a person sees
 * in the sandbox agrees with the number on the scorecard behind it. Opening on
 * round defaults would make them disagree immediately and silently.
 */
export function dcfBaselines(dcfSummary) {
  const a = dcfSummary?.assumptions;
  // Clamped here rather than only at the presets, so the baseline the sliders
  // show and the baseline the presets are derived from are the same number.
  // Unclamped, a company with 90% filed growth showed a slider pinned at 45
  // while its bear case was computed from 90 — a bear case bearing no visible
  // relation to the base it claimed to be worse than.
  return clampAssumptions({
    growthPct: Math.round((a?.growthRate ?? 0.06) * 100),
    multiple: Math.round(a?.terminalMultiple ?? 15),
    discountPct: Math.round((a?.discountRate ?? 0.095) * 1000) / 10
  });
}

/**
 * Bear, base and bull, relative to the company's own baseline.
 *
 * The bear case is 35% *worse* rather than 35% lower, which is not the same
 * thing for a shrinking business: a company already contracting has a bear case
 * of faster decline, not slower expansion. Multiplying a negative growth rate
 * by 0.65 would make the bear case optimistic.
 */
export function presetAssumptions(preset, baselines) {
  const g = baselines.growthPct;
  if (preset === 'bear') {
    return clampAssumptions({
      growthPct: Math.round(g >= 0 ? g * 0.65 : g * 1.35),
      multiple: 16,
      discountPct: 11.0
    });
  }
  if (preset === 'bull') {
    return clampAssumptions({
      growthPct: Math.round(g >= 0 ? g * 1.3 : g * 0.7),
      multiple: 32,
      discountPct: 9.0
    });
  }
  return clampAssumptions(baselines);
}

/**
 * Hold every assumption inside the slider's range.
 *
 * Clamped at both ends, which the original did not do: it capped the bear case
 * from below and the bull case from above, each leaving the other end open. A
 * company whose filed growth sits outside the range — a steep decline, or a
 * recovery year off a small base — then produced a preset the slider could not
 * represent, so the control clamped silently while the fair value beneath it
 * was computed from the unclamped figure. The two disagreed, and nothing said so.
 *
 * Baselines go through here too, for the same reason: they come from the
 * filings, and the filings are not obliged to land inside a UI range.
 */
export function clampAssumptions(a) {
  const clamp = (v, { min, max }) => Math.min(max, Math.max(min, v));
  return {
    growthPct: clamp(a.growthPct, LIMITS.growthPct),
    multiple: clamp(a.multiple, LIMITS.multiple),
    discountPct: clamp(a.discountPct, LIMITS.discountPct)
  };
}

/**
 * Why the model cannot be run, or null if it can.
 *
 * Returned as a reason rather than a boolean because each of these needs a
 * different sentence. The previous build substituted $1bn of free cash flow
 * where it was missing and produced a fair value for companies that were
 * burning cash — a confident number for a question the filings do not answer.
 */
export function dcfBlockedReason({ dcfSummary, cashFlowBase, shares }) {
  if (dcfSummary?.applicable === false) return dcfSummary.reason || 'not-applicable';
  if (!isFinite(cashFlowBase) || cashFlowBase <= 0) return 'negative-fcf';
  if (!isFinite(shares) || shares <= 0) return 'no-share-count';
  return null;
}

export const BLOCKED_EXPLANATIONS = {
  'negative-fcf':
    'This company is not generating positive free cash flow, so a discounted cash flow ' +
    'model has nothing to discount. Judge it on the balance sheet and the path back to ' +
    'cash generation instead.',
  'no-share-count':
    'The diluted share count is not in the filings for this listing, so a per-share ' +
    'value cannot be derived.',
  'not-meaningful-for-financials':
    'Free cash flow is not owner earnings for a bank or insurer — deposit and loan flows ' +
    'dominate it. Book value and return on equity are the measures that apply here.'
};

export function explainBlocked(reason) {
  return BLOCKED_EXPLANATIONS[reason] || 'This model cannot be run on the available filings.';
}

/**
 * Two-stage DCF: five explicit years, then a terminal exit multiple.
 *
 * @returns per-year rows, the discounted components, and the fair value.
 */
export function projectDcf({ cashFlowBase, shares, netCash = 0, growthPct, multiple, discountPct }) {
  const g = growthPct / 100;
  const r = discountPct / 100;

  const rows = [];
  let fcf = cashFlowBase;
  let cumulativePV = 0;

  for (let t = 1; t <= PROJECTION_YEARS; t++) {
    fcf = fcf * (1 + g);
    const pv = fcf / Math.pow(1 + r, t);
    cumulativePV += pv;
    rows.push({ year: t, fcf, pv });
  }

  const terminalValue = fcf * multiple;
  const pvTerminal = terminalValue / Math.pow(1 + r, PROJECTION_YEARS);
  const enterpriseValue = cumulativePV + pvTerminal;
  const equityValue = enterpriseValue + netCash;

  return {
    rows,
    cumulativePV,
    terminalValue,
    pvTerminal,
    enterpriseValue,
    equityValue,
    fairValue: equityValue / shares
  };
}

/**
 * What the result means against the traded price.
 *
 * Four outcomes, and three of them exist because the obvious single number is
 * misleading somewhere:
 *
 * - `no-equity-value` — the discounted flows do not cover the debt.
 * - `divergent` — a fair value several multiples from the price almost always
 *   means the assumptions are wrong, or the market is pricing something the
 *   filings do not show. Presenting that as an enormous margin of safety
 *   invites exactly the wrong conclusion.
 * - `undervalued` — a genuine margin of safety.
 * - `overvalued` — stated as a *premium over* fair value rather than a negative
 *   margin of safety, because the margin form reaches −188% on an expensive
 *   stock and stops carrying meaning.
 */
export function dcfVerdict(fairValue, price) {
  if (!isFinite(fairValue) || fairValue <= 0) {
    return { kind: 'no-equity-value', pct: null, factor: null };
  }
  if (!isFinite(price) || price <= 0) {
    return { kind: 'no-price', pct: null, factor: null };
  }

  const factor = fairValue / price;
  if (factor >= 3 || factor <= 1 / 3) {
    return { kind: 'divergent', pct: null, factor };
  }
  if (fairValue > price) {
    return { kind: 'undervalued', pct: ((fairValue - price) / fairValue) * 100, factor };
  }
  return { kind: 'overvalued', pct: ((price - fairValue) / fairValue) * 100, factor };
}
